clients/drcachesim/tests/invariant_checker_test.cpp

```diff
@@ -345,6 +345,142 @@ check_sane_control_flow()
+            gen_instr(TID, 1),
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, 3),
+        };
+        if (!run_checker(
+                memrefs, true,
+                { "Non-explicit control flow has no marker @ kernel_event marker", TID,
+                  /*ref_ordinal=*/2, /*last_timestamp=*/0,
+                  /*instrs_since_last_timestamp=*/1 },
+                "Failed to catch PC discontinuity for an instruction followed by "
+                "kernel xfer marker")) {
+            return false;
+        }
+    }
+    // Correct test: Transition from instr to kernel_xfer event marker. goes to the next
+    // instruction.
+    {
+        std::vector<memref_t> memrefs = {
+            gen_instr(TID, 1),   gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, 2),
+            gen_instr(TID, 101), gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_XFER, 102),
+            gen_instr(TID, 2),
+        };
+        if (!run_checker(memrefs, false)) {
+            return false;
+        }
+    }
+    // Correct test: We should skip the check if there is no instruction before the
+    // kernel event.
+    {
+        std::vector<memref_t> memrefs = {
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, 3),
+        };
+        if (!run_checker(memrefs, false)) {
+            return false;
+        }
+    }
+    // Correct test: Pre-signal instr continues after signal.
+    {
+        std::vector<memref_t> memrefs = {
+            gen_instr(TID, 2),   gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, 2),
+            gen_instr(TID, 101), gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_XFER, 102),
+            gen_instr(TID, 2),
+        };
+        if (!run_checker(memrefs, false)) {
+            return false;
+        }
+    }
+    // Correct test: We should not report a PC discontinuity when the previous instr is
+    // of type TRACE_TYPE_INSTR_SYSENTER.
+    {
+        std::vector<memref_t> memrefs = {
+            gen_instr(TID, 5),
+            gen_instr_type(TRACE_TYPE_INSTR_SYSENTER, TID, 6),
+            gen_marker(TID, TRACE_MARKER_TYPE_TIMESTAMP, 2),
+            gen_marker(TID, TRACE_MARKER_TYPE_CPU_ID, 3),
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, 2),
+            gen_instr(TID, 101),
+        };
+        if (!run_checker(memrefs, false)) {
+            return false;
+        }
+    }
+    // Correct test: RSEQ abort in last signal context.
+    {
+        std::vector<memref_t> memrefs = {
+            gen_instr(TID, 1),
+            // The RSEQ_ABORT marker is always follwed by a KERNEL_EVENT marker.
+            gen_marker(TID, TRACE_MARKER_TYPE_RSEQ_ABORT, 40),
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, 40),
+            // We get a signal after the RSEQ abort.
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, 4),
+        };
+        if (!run_checker(memrefs, false)) {
+            return false;
+        }
+    }
+    // Correct test: Branch before signal.
```

Comment:
Add here that this is correct only because the branch doesn't have an encoding with it.